File: AutomaticTimetableGeneratorSystem-main/timetable-frontend/src/services/subjectService.js
import api from "./api";

const BASE_URL = "/subjects";

const subjectService = {

  // Get All Subjects
  getAllSubjects: async () => {

    const response = await api.get(BASE_URL);

    return response.data;

  },

  // Get Subject By Id
  getSubjectById: async (id) => {

    const response = await api.get(`${BASE_URL}/${id}`);

    return response.data;

  },

  // Create Subject
  createSubject: async (subject) => {

    const response = await api.post(
      BASE_URL,
      subject
    );

    return response.data;

  },

  // Update Subject
  updateSubject: async (id, subject) => {

    const response = await api.put(
      `${BASE_URL}/${id}`,
      subject
    );

    return response.data;

  },

  // Delete Subject
  deleteSubject: async (id) => {

    const response = await api.delete(
      `${BASE_URL}/${id}`
    );

    return response.data;

  },

  // Get Subject By Code
  getSubjectByCode: async (subjectCode) => {

    const response = await api.get(
      `${BASE_URL}/code/${subjectCode}`
    );

    return response.data;

  },

  // Get Subjects By Department
  getSubjectsByDepartment: async (departmentId) => {

    const response = await api.get(
      `${BASE_URL}/department/${departmentId}`
    );

    return response.data;

  },

  // Get Subjects By Course
  getSubjectsByCourse: async (courseId) => {

    const response = await api.get(
      `${BASE_URL}/course/${courseId}`
    );

    return response.data;

  },

  // Get Subjects By Semester
  getSubjectsBySemester: async (semesterId) => {

    const response = await api.get(
      `${BASE_URL}/semester/${semesterId}`
    );

    return response.data;

  },

  // Get Subjects By Faculty
  getSubjectsByFaculty: async (facultyId) => {

    const response = await api.get(
      `${BASE_URL}/faculty/${facultyId}`
    );

    return response.data;

  }

};

export default subjectService;

File: AutomaticTimetableGeneratorSystem-main/timetable-backend/src/main/java/com/timetable/service/SubjectService.java
package com.timetable.service;

import java.util.List;

import com.timetable.dto.subject.SubjectRequest;
import com.timetable.dto.subject.SubjectResponse;

public interface SubjectService {

    /**
     * Save Subject
     */
    SubjectResponse saveSubject(SubjectRequest request);

    /**
     * Update Subject
     */
    SubjectResponse updateSubject(Long subjectId, SubjectRequest request);

    /**
     * Delete Subject
     */
    void deleteSubject(Long subjectId);

    /**
     * Get Subject By ID
     */
    SubjectResponse getSubjectById(Long subjectId);

    /**
     * Get All Subjects
     */
    List<SubjectResponse> getAllSubjects();

    /**
     * Get Subject By Subject Code
     */
    SubjectResponse getSubjectByCode(String subjectCode);

    /**
     * Get Subjects By Department
     */
    List<SubjectResponse> getSubjectsByDepartment(Long departmentId);

    /**
     * Get Subjects By Course
     */
    List<SubjectResponse> getSubjectsByCourse(Long courseId);

    /**
     * Get Subjects By Semester
     */
    List<SubjectResponse> getSubjectsBySemester(Long semesterId);

    /**
     * Get Subjects By Faculty
     */
    List<SubjectResponse> getSubjectsByFaculty(Long facultyId);
}